quickwit/quickwit-serve/src/ingest_api/response.rs

```diff
@@ -0,0 +1,258 @@
+    /// Number of docs successfully ingested
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub num_ingested_docs: Option<u64>,
+    /// Number of docs rejected because of parsing errors
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub num_rejected_docs: Option<u64>,
+    /// Detailed description of parsing errors (available if the path param
+    /// `detailed_parse_failures` is set to `true`)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub parse_failures: Option<Vec<RestParseFailure>>,
+}
+
+impl RestIngestResponse {
+    pub(crate) fn from_ingest_v1(ingest_response: IngestResponse) -> Self {
+        Self {
+            num_docs_for_processing: ingest_response.num_docs_for_processing,
+            num_ingested_docs: None,
+            num_rejected_docs: None,
+            parse_failures: None,
+        }
+    }
+
+    /// Generate a detailed failure description if `doc_batch_clone_opt.is_some()`
+    pub(crate) fn from_ingest_v2(
+        mut ingest_response: IngestResponseV2,
+        doc_batch_clone_opt: Option<DocBatchV2>,
+        num_docs_for_processing: u64,
+    ) -> Result<Self, IngestServiceError> {
+        let num_responses = ingest_response.successes.len() + ingest_response.failures.len();
+        if num_responses != 1 {
+            return Err(IngestServiceError::Internal(format!(
+                "expected a single failure/success, got {num_responses}",
+            )));
+        }
+        if let Some(failure_resp) = ingest_response.failures.pop() {
+            return Err(failure_resp.into());
+        }
+        let success_resp = ingest_response.successes.pop().unwrap();
+
+        let mut resp = Self {
+            num_docs_for_processing,
+            num_ingested_docs: Some(success_resp.num_ingested_docs as u64),
+            num_rejected_docs: Some(success_resp.parse_failures.len() as u64),
+            parse_failures: None,
+        };
+        if let Some(doc_batch) = doc_batch_clone_opt {
+            let docs: BTreeMap<DocUid, Bytes> = doc_batch.into_docs().collect();
+            let mut parse_failures = Vec::with_capacity(success_resp.parse_failures.len());
+            for failure in success_resp.parse_failures {
+                let doc = docs.get(&failure.doc_uid()).ok_or_else(|| {
+                    IngestServiceError::Internal(format!(
+                        "failed doc_uid {} not found in the original doc batch",
+                        failure.doc_uid()
+                    ))
+                })?;
+                parse_failures.push(RestParseFailure {
+                    message: failure.message,
+                    document: String::from_utf8(doc.to_vec()).unwrap(),
+                    reason: ParseFailureReason::from_i32(failure.reason).unwrap_or_default(),
+                });
+            }
+            resp.parse_failures = Some(parse_failures);
+        }
+        Ok(resp)
+    }
+}
+
+impl Add for RestIngestResponse {
```

Comment:
Ok, this explains the `response + response` seen earlier. Isn't this more readable and explicit?

```rust
impl RestIngestResponse {
  /// Merges the `other` response into `self`.
  fn merge(&mut self, response: Self) {
    ...
  }
}
```

Reply:
agreed, I usually don't like overloading operators, I don't what's gotten into me :-\